### DISCUSSION TITLE
Beginner's publish question

### DISCUSSION THREAD

**Discussion by smnrgrs** [eclipse-zenoh/roadmap#153] [Q&A] [answered]
The simple examples tend to create a publisher (z_declare_publisher) with a keyexpr, and then call z_publisher_put when they want to publish updates with that key (C API).

I have 150 sensors, which will each publish an update every second. As I see it my options are:

1. Don't create publishers, just call z_put for each sensor update, specifying the serial number of the sensor at the end of the keyexpr.
2. Create a publisher up front for each sensor when they connect to the system, specifying the serial number of the sensor at the end of the keyexpr. Reuse the publisher every time a new data update is published by the sensor.
3. Can I reuse a publisher to publish multiple keys? I can't see how in the C API?

 I wonder what the pros and cons of these might be, and which would be viewed as the way Zenoh was designed to be used? Which is more likely to scale to larger networks of sensors?

Thanks

**Comment by gabrik** [comment, marked as answer]
Hi @smnrgrs,

TL;DR;

1 and 2 ) Using the publisher allows for some wire optimizations that are not present when using the `z_put` directly.
So if the key expression being used by the sensor is always the same and it is known a priori I advocate for creating the publishers.

3) I think the publisher is tied to the key expression used in the declaration, this is true also in other languages not only in C.
In general, if the amount of key expression being published by every sensor is small and known a priori you can declare a publisher per key expression.
